File: api-java/src/main/java/fr/huiitre/tools/modules/riot/valorant/api/ValorantAuthController.java
package fr.huiitre.tools.modules.riot.valorant.api;

import fr.huiitre.tools.modules.core.common.api.RequiredRole;
import fr.huiitre.tools.modules.core.role.domain.RoleCode;
import fr.huiitre.tools.modules.riot.valorant.application.core.command.LinkValorantAccountCommand;
import fr.huiitre.tools.modules.riot.valorant.application.core.command.RenameValorantAccountCommand;
import fr.huiitre.tools.modules.riot.valorant.application.core.usecase.GetValorantAccessTokenUseCase;
import fr.huiitre.tools.modules.riot.valorant.application.core.usecase.LinkValorantAccountUseCase;
import fr.huiitre.tools.modules.riot.valorant.application.core.usecase.ListValorantAccountsUseCase;
import fr.huiitre.tools.modules.riot.valorant.application.core.usecase.RenameValorantAccountUseCase;
import fr.huiitre.tools.modules.riot.valorant.application.core.usecase.UnlinkValorantAccountUseCase;
import fr.huiitre.tools.modules.riot.valorant.application.core.view.ValorantAccountAuthView;
import fr.huiitre.tools.modules.riot.valorant.application.core.view.ValorantAccountView;
import fr.huiitre.tools.modules.riot.valorant.application.core.view.ValorantTokenView;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/riot/valorant/accounts")
public class ValorantAuthController {

    private final LinkValorantAccountUseCase linkValorantAccountUseCase;
    private final ListValorantAccountsUseCase listValorantAccountsUseCase;
    private final UnlinkValorantAccountUseCase unlinkValorantAccountUseCase;
    private final RenameValorantAccountUseCase renameValorantAccountUseCase;
    private final GetValorantAccessTokenUseCase getValorantAccessTokenUseCase;

    public ValorantAuthController(LinkValorantAccountUseCase linkValorantAccountUseCase,
                                   ListValorantAccountsUseCase listValorantAccountsUseCase,
                                   UnlinkValorantAccountUseCase unlinkValorantAccountUseCase,
                                   RenameValorantAccountUseCase renameValorantAccountUseCase,
                                   GetValorantAccessTokenUseCase getValorantAccessTokenUseCase) {
        this.linkValorantAccountUseCase = linkValorantAccountUseCase;
        this.listValorantAccountsUseCase = listValorantAccountsUseCase;
        this.unlinkValorantAccountUseCase = unlinkValorantAccountUseCase;
        this.renameValorantAccountUseCase = renameValorantAccountUseCase;
        this.getValorantAccessTokenUseCase = getValorantAccessTokenUseCase;
    }

    @RequiredRole(RoleCode.USER)
    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    public ValorantAccountAuthView linkAccount(@RequestBody LinkValorantAccountCommand command) {
        return linkValorantAccountUseCase.execute(command);
    }

    @RequiredRole(RoleCode.READ_ONLY)
    @GetMapping
    public List<ValorantAccountView> listAccounts() {
        return listValorantAccountsUseCase.execute();
    }

    @RequiredRole(RoleCode.READ_ONLY)
    @GetMapping("/{accountId}/refresh")
    public ValorantTokenView refresh(@PathVariable Long accountId) {
        return getValorantAccessTokenUseCase.execute(accountId);
    }

    @RequiredRole(RoleCode.USER)
    @DeleteMapping("/{accountId}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public void unlinkAccount(@PathVariable Long accountId) {
        unlinkValorantAccountUseCase.execute(accountId);
    }

    @RequiredRole(RoleCode.USER)
    @PutMapping("/{accountId}")
    public ValorantAccountView renameAccount(@PathVariable Long accountId, @RequestBody RenameValorantAccountCommand command) {
        return renameValorantAccountUseCase.execute(accountId, command);
    }
}

File: api/Modules/Riot/Valorant/Api/ValorantAuthController.cs
using Microsoft.AspNetCore.Mvc;
using Tools.Api.Modules.Riot.Valorant.Application.Core.Commands;
using Tools.Api.Modules.Riot.Valorant.Application.Core.Usecases;
using Tools.Api.Modules.Riot.Valorant.Application.Core.Views;

namespace Tools.Api.Modules.Riot.Valorant.Api;

// Comptes Valorant liés à un compte Tools.
//
// Les use cases sont résolus par action ([FromServices]) : un use case sécurisé applique son
// contrôle dès sa construction, une action ne doit donc construire que celui dont elle se sert.
// Aucun attribut d'autorisation ici, le rôle exigé est porté par le use case.
[ApiController]
[Route("riot/valorant/accounts")]
public class ValorantAuthController : ControllerBase
{
    [HttpPost]
    public async Task<IActionResult> LinkAccount(
        [FromServices] LinkValorantAccountUseCase linkValorantAccountUseCase,
        [FromBody] LinkValorantAccountCommand command
    )
    {
        var account = await linkValorantAccountUseCase.Execute(command);

        return StatusCode(StatusCodes.Status201Created, account);
    }

    [HttpGet]
    public Task<List<ValorantAccountView>> ListAccounts(
        [FromServices] ListValorantAccountsUseCase listValorantAccountsUseCase
    )
    {
        return listValorantAccountsUseCase.Execute();
    }

    [HttpGet("{accountId:long}/refresh")]
    public Task<ValorantTokenView> Refresh(
        [FromServices] GetValorantAccessTokenUseCase getValorantAccessTokenUseCase,
        [FromRoute] long accountId
    )
    {
        return getValorantAccessTokenUseCase.Execute(accountId);
    }

    [HttpDelete("{accountId:long}")]
    public async Task<IActionResult> UnlinkAccount(
        [FromServices] UnlinkValorantAccountUseCase unlinkValorantAccountUseCase,
        [FromRoute] long accountId
    )
    {
        await unlinkValorantAccountUseCase.Execute(accountId);

        return NoContent();
    }

    [HttpPut("{accountId:long}")]
    public Task<ValorantAccountView> RenameAccount(
        [FromServices] RenameValorantAccountUseCase renameValorantAccountUseCase,
        [FromRoute] long accountId,
        [FromBody] RenameValorantAccountCommand command
    )
    {
        return renameValorantAccountUseCase.Execute(accountId, command);
    }
}
